File: Freedom/微信/基本类库/Common/WXInfoViewController.m
//  TLInfoViewController.m
//  Freedom
// Created by Super
#import "WXInfoViewController.h"
#import <BlocksKit/BlocksKit+UIKit.h>
@interface WXInfoCell ()
@property (nonatomic, strong) UILabel *subTitleLabel;
@end
@implementation WXInfoCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self.textLabel setFont:[UIFont systemFontOfSize:15.0f]];
        [self.contentView addSubview:self.subTitleLabel];
        [self p_addMasonry];
    }
    return self;
}
- (void)setInfo:(WXInfo *)info{
    _info = info;
    [self.textLabel setText:info.title];
    [self.subTitleLabel setText:info.subTitle];
    [self setAccessoryType:info.showDisclosureIndicator ? UITableViewCellAccessoryDisclosureIndicator : UITableViewCellAccessoryNone];
    [self setSelectionStyle:info.disableHighlight ? UITableViewCellSelectionStyleNone : UITableViewCellSelectionStyleDefault];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.subTitleLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.mas_equalTo(self.contentView);
        make.left.mas_equalTo(self.contentView).mas_offset(APPW * 0.28);
        make.right.mas_lessThanOrEqualTo(self.contentView);
    }];
}
#pragma mark - Getter -
- (UILabel *)subTitleLabel{
    if (_subTitleLabel == nil) {
        _subTitleLabel = [[UILabel alloc] init];
        [_subTitleLabel setFont:[UIFont systemFontOfSize:15.0f]];
    }
    return _subTitleLabel;
}
@end
@interface WXInfoButtonCell ()
@property (nonatomic, strong) UIButton *button;
@end
@implementation WXInfoButtonCell
- (id)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithStyle:style reuseIdentifier:reuseIdentifier]) {
        [self setBackgroundColor:[UIColor clearColor]];
        [self setSelectionStyle:UITableViewCellSelectionStyleNone];
        [self setAccessoryType:UITableViewCellAccessoryNone];
        [self setSeparatorInset:UIEdgeInsetsMake(0, APPW / 2, 0, APPW / 2)];
        [self setLayoutMargins:UIEdgeInsetsMake(0, APPW / 2, 0, APPW / 2)];
        [self.contentView addSubview:self.button];
        [self p_addMasonry];
    }
    return self;
}
- (void)setInfo:(WXInfo *)info{
    _info = info;
    [self.button setTitle:info.title forState:UIControlStateNormal];
    [self.button setBackgroundColor:info.buttonColor];
    [self.button setBackgroundImage:[FreedomTools imageWithColor:info.buttonHLColor] forState:UIControlStateHighlighted];
    [self.button setTitleColor:info.titleColor forState:UIControlStateNormal];
    [self.button.layer setBorderColor:info.buttonBorderColor.CGColor];
}
#pragma mark - Event Response -
- (void)cellButtonDown:(UIButton *)sender{
    if (_delegate && [_delegate respondsToSelector:@selector(infoButtonCellClicked:)]) {
        [_delegate infoButtonCellClicked:self.info];
    }
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.button mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.and.top.mas_equalTo(self.contentView);
        make.width.mas_equalTo(self.contentView).multipliedBy(0.92);
        make.height.mas_equalTo(self.contentView).multipliedBy(0.78);
    }];
}
#pragma mark - Getter -
- (UIButton *)button{
    if (_button == nil) {
        _button = [[UIButton alloc] init];
        [_button.layer setMasksToBounds:YES];
        [_button.layer setCornerRadius:4.0f];
        [_button.layer setBorderWidth:1];
        [_button.titleLabel setFont:[UIFont systemFontOfSize:16.0f]];
        [_button addTarget:self action:@selector(cellButtonDown:) forControlEvents:UIControlEventTouchUpInside];
    }
    return _button;
}
@end
@implementation WXInfoHeaderFooterView
- (id) initWithReuseIdentifier:(NSString *)reuseIdentifier{
    if (self = [super initWithReuseIdentifier:reuseIdentifier]) {
        [self.contentView setBackgroundColor:[UIColor lightGrayColor]];
    }
    return self;
}
@end
@implementation WXInfoViewController
- (void) loadView{
    self.view = [[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, APPH)];
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStyleGrouped];
    [self.tableView setTableHeaderView:[[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, 15.0f)]];
    [self.tableView setTableFooterView:[[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, 12.0f)]];
    [self.tableView setBackgroundColor:[UIColor lightGrayColor]];
}
- (void)viewDidLoad{
    [super viewDidLoad];
    [self.tableView setTableHeaderView:[[UIView alloc] initWithFrame:CGRectMake(0, 0, APPW, 10.0f)]];
    [self.tableView registerClass:[WXInfoHeaderFooterView class] forHeaderFooterViewReuseIdentifier:@"TLInfoHeaderFooterView"];
    [self.tableView registerClass:[WXInfoCell class] forCellReuseIdentifier:@"TLInfoCell"];
    [self.tableView registerClass:[WXInfoButtonCell class] forCellReuseIdentifier:@"TLInfoButtonCell"];
}
- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    [self.tableView setSeparatorStyle:UITableViewCellSeparatorStyleNone];
}
- (void)dealloc{
#ifdef DEBUG_MEMERY
    DLog(@"dealloc %@", self.navigationItem.title);
#endif
}
#pragma mark - Delegate -
//MARK: UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return self.data.count;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    NSArray *temp = self.data[section];
    return [temp count];
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    NSArray *group = self.data[indexPath.section];
    WXInfo *info = [group objectAtIndex:indexPath.row];
    id cell;
    if (info.type == TLInfoTypeButton) {
        cell = [tableView dequeueReusableCellWithIdentifier:@"TLInfoButtonCell"];
        [cell setDelegate:self];
    }else{
        cell = [tableView dequeueReusableCellWithIdentifier:@"TLInfoCell"];
    }
    [cell setInfo:info];
    
    if (indexPath.row == 0 && info.type != TLInfoTypeButton) {
        [cell setTopLineStyle:TLCellLineStyleFill];
    }else{
        [cell setTopLineStyle:TLCellLineStyleNone];
    }
    if (info.type == TLInfoTypeButton) {
        [cell setBottomLineStyle:TLCellLineStyleNone];
    }else if (indexPath.row == group.count - 1) {
        [cell setBottomLineStyle:TLCellLineStyleFill];
    }else{
        [cell setBottomLineStyle:TLCellLineStyleDefault];
    }
    return cell;
}
- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section{
    return [tableView dequeueReusableHeaderFooterViewWithIdentifier:@"TLInfoHeaderFooterView"];
}
- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section{
    return [tableView dequeueReusableHeaderFooterViewWithIdentifier:@"TLInfoHeaderFooterView"];
}
//MARK: TLTableViewDelegate
- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    [tableView deselectRowAtIndexPath:indexPath animated:NO];
}
- (CGFloat)tableView:(UITableView *)tableView heightForHeaderInSection:(NSInteger)section{
    return 5.0f;
}
- (CGFloat)tableView:(UITableView *)tableView heightForFooterInSection:(NSInteger)section{
    return 15.0f;
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    WXInfo *info = [self.data[indexPath.section] objectAtIndex:indexPath.row];
    if (info.type == TLInfoTypeButton) {
        return 50.0f;
    }
    return 44.0f;
}
//MARK: TLInfoButtonCellDelegate
- (void)infoButtonCellClicked:(WXInfo *)info{
    [self showAlerWithtitle:@"子类未处理按钮点击事件" message:[NSString stringWithFormat:@"Title: %@", info.title] style:UIAlertControllerStyleAlert ac1:^UIAlertAction *{
        return [UIAlertAction actionWithTitle:@"取消" style:UIAlertActionStyleCancel handler:^(UIAlertAction * _Nonnull action) {}];
    } ac2:nil ac3:nil completion:nil];
}
@end

File: Freedom/微信/基本类库/Common/WXInfoViewController.swift
//
//  WXInfoViewController.swift
//  Freedom
import SnapKit
import XExtension
import Foundation
protocol WXInfoButtonCellDelegate: NSObjectProtocol {
    func infoButtonCellClicked(_ info: WXInfo)
}
class WXInfoButtonCell: WXTableViewCell {
    weak var delegate: WXInfoButtonCellDelegate?
    var info: WXInfo = WXInfo() {
        didSet {
            button.setTitle(info.title, for: .normal)
            button.backgroundColor = info.buttonColor
            button.setBackgroundImage(UIImage.imageWithColor(info.buttonHLColor), for: .highlighted)
            button.setTitleColor(info.titleColor, for: .normal)
            button.layer.borderColor = info.buttonBorderColor.cgColor
        }
    }
    private lazy var button: UIButton =  {
        let button = UIButton()
        button.layer.masksToBounds = true
        button.layer.cornerRadius = 4.0
        button.layer.borderWidth = 1
        button.titleLabel?.font = UIFont.systemFont(ofSize: 16.0)
        button.addTarget(self, action: #selector(self.cellButtonDown(_:)), for: .touchUpInside)
        return button
    }()
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        backgroundColor = UIColor.clear
        selectionStyle = .none
        accessoryType = .none
        separatorInset = UIEdgeInsets(top: 0, left: APPW / 2, bottom: 0, right: APPW / 2)
        layoutMargins = UIEdgeInsets(top: 0, left: APPW / 2, bottom: 0, right: APPW / 2)
        contentView.addSubview(button)
        button.snp.makeConstraints { (make) in
            make.centerX.top.equalTo(self.contentView)
            make.width.equalTo(self.contentView).multipliedBy(0.92)
            make.height.equalTo(self.contentView).multipliedBy(0.78)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    @objc func cellButtonDown(_ sender: UIButton) {
        delegate?.infoButtonCellClicked(info)
    }
}

class WXInfoCell: WXTableViewCell {
    private var subTitleLabel = UILabel()
    var info: WXInfo = WXInfo() {
        didSet {
            textLabel?.text = info.title
            subTitleLabel.text = info.subTitle
            accessoryType = info.showDisclosureIndicator ? .disclosureIndicator : .none
            selectionStyle = info.disableHighlight ? .none : .default
        }
    }
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        subTitleLabel.font = UIFont.systemFont(ofSize: 15.0)
        textLabel?.font = UIFont.systemFont(ofSize: 15.0)
        contentView.addSubview(subTitleLabel)
        subTitleLabel.snp.makeConstraints { (make) in
            make.centerY.equalTo(self.contentView)
            make.left.equalTo(self.contentView).offset(APPW * 0.28)
            make.right.lessThanOrEqualTo(self.contentView)
        }
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class WXInfoHeaderFooterView: UITableViewHeaderFooterView {
    override init(reuseIdentifier: String?) {
        super.init(reuseIdentifier: reuseIdentifier)
        contentView.backgroundColor = UIColor.lightGray
    }
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

class WXInfoViewController: BaseTableViewController, WXInfoButtonCellDelegate {
    var data: [[WXInfo]] = []
    override func loadView() {
        view = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: APPH))
        tableView = UITableView(frame: view.bounds, style: .grouped)
        tableView.tableHeaderView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: 15.0))
        tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: 12.0))
        tableView.backgroundColor = UIColor.lightGray
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.tableHeaderView = UIView(frame: CGRect(x: 0, y: 0, width: APPW, height: 10.0))
        tableView.register(WXInfoHeaderFooterView.self, forHeaderFooterViewReuseIdentifier: "TLInfoHeaderFooterView")
        tableView.register(WXInfoCell.self, forCellReuseIdentifier: "TLInfoCell")
        tableView.register(WXInfoButtonCell.self, forCellReuseIdentifier: "TLInfoButtonCell")
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        tableView.separatorStyle = .none
    }
    override func numberOfSections(in tableView: UITableView) -> Int {
        return data.count
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        let temp = data[section]
        return temp.count
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let group = data[indexPath.section]
        let info = group[indexPath.row]
        var cell: WXTableViewCell
        if info.type == TLInfoType.button.rawValue {
            cell = tableView.dequeueReusableCell(withIdentifier: WXInfoButtonCell.identifier) as! WXInfoButtonCell
            (cell as! WXInfoButtonCell).delegate = self
            (cell as! WXInfoButtonCell).info = info
        } else {
            cell = tableView.dequeueReusableCell(withIdentifier: WXInfoCell.identifier) as! WXTableViewCell
            (cell as! WXInfoCell).info = info
        }
        if indexPath.row == 0 && info.type != TLInfoType.button.rawValue {
            cell.topLineStyle = .fill
        } else {
            cell.topLineStyle = .none
        }
        if info.type == TLInfoType.button.rawValue {
            cell.bottomLineStyle = .none
        } else if indexPath.row == group.count - 1 {
            cell.bottomLineStyle = .fill
        } else {
            cell.bottomLineStyle = .default
        }
        return cell
    }
    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        return tableView.dequeueReusableHeaderFooterView(withIdentifier: "TLInfoHeaderFooterView")
    }
    override func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        return tableView.dequeueReusableHeaderFooterView(withIdentifier: "TLInfoHeaderFooterView")
    }
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: false)
    }
    override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        return 5.0
    }
    override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        return 15.0
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let info = data[indexPath.section][indexPath.row]
        if info.type == Int32(TLInfoType.button.rawValue) {
            return 50.0
        }
        return 44.0
    }
    func infoButtonCellClicked(_ info: WXInfo) {
        let alert = UIAlertController("子类未处理按钮点击事件", "", T1: "取消", T2: "", confirm1: {

        }, confirm2: {

        })
        self.present(alert, animated: true, completion: nil)
    }
}
